src/library/scala/collection/Iterable.scala

```diff
@@ -549,21 +549,22 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     *
     */
   def groupBy[K](f: A => K): immutable.Map[K, C] = {
-    val m = mutable.Map.empty[K, Builder[A, C]]
+    val m = mutable.HashMap.empty[K, Any /*C | Builder[A, C] */]
     val it = iterator
-    while (it.hasNext) {
-      val elem = it.next()
-      val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newSpecificBuilder)
-      bldr += elem
-    }
-    var result = immutable.HashMap.empty[K, C]
-    val mapIt = m.iterator
-    while (mapIt.hasNext) {
-      val (k, v) = mapIt.next()
-      result = result.updated(k, v.result())
+    if (it.isEmpty) immutable.Map.empty
+    else {
+      while (it.hasNext) {
+        val elem = it.next()
+        val key  = f(elem)
+        val bldr = m.getOrElseUpdate(key, newSpecificBuilder).asInstanceOf[Builder[A, C]]
+        bldr += elem
+      }
+      m.mapValuesInPlace((_, v) => v.asInstanceOf[Builder[A, C]].result())
```

Comment:
there is a special case here. If all value map to the same key then `this == m.head._2.result()` but probably would not be `eq`, so to help structural sharing and reduce garbage we could do 
``` 
if (m.size == 1) new Map.Map1(m.head._1, this) //if m.head._1 is the cheapest way to get the key
    else ...```

Reply:
We would need to limit such an optimizations cases where `this` is an immutable collection and we can reason that `val b = this.newSpecicicBuilder; b.addAll(this); val res = b.result(); assert(res.sameElements(b) && res.getClass == this.getClass)`.

Reply:
It would be nice in general to know is `this.isImmutable`